mono/mini/mini-llvm.c

```diff
@@ -7341,6 +7341,27 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			break;
 		}
 		case OP_XEQUAL: {
+#if defined(TARGET_AMD64) || defined(TARGET_X86)
+			int nelemsi8 = 16; // TODO: calculate vector width in bytes for current CPU
```

Comment:
LLVM is supposed to convert the code we generate to this intrinsic.

Reply:
but it doesn't? 🙂 . I ported this code https://godbolt.org/z/-GEmBj (after opts: https://godbolt.org/z/dIzTDE)

Reply:
I.e.:
https://github.com/llvm-mirror/llvm/blob/master/test/CodeGen/X86/vector-compare-all_of.ll#L35

Reply:
@vargaz hm.. comments state that it should generate movmskpd but it doesn't. Instead it generate something similar to what AVX-512 should generate (but my CPU doesn't have it). Let me benchmark both my and the current impl.

Reply:
hm.. will be fixed once we move to LLVM 9 ? https://godbolt.org/z/ITU_2g

Reply:
It generated that when I wrote that code, maybe something changed in the meantime.

Reply:
I think it was generating it on my machine too...

Reply:
my Fix makes `==` 50% faster according to some benchmarks. Here is the IR we emit (unoptimized):
```llvm
source_filename = "jit-module-Foo"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define i8 @"HelloWorld.Program:Foo (System.Numerics.Vector`1<long>,System.Numerics.Vector`1<long>)"(i64 %arg_vec1, i64, i64 %arg_vec2, i64) local_unnamed_addr #0 {
BB0:
  %.0.vec.insert = insertelement <2 x i64> undef, i64 %arg_vec1, i32 0
  %.8.vec.insert = insertelement <2 x i64> %.0.vec.insert, i64 %0, i32 1
  %.0.vec.insert4 = insertelement <2 x i64> undef, i64 %arg_vec2, i32 0
  %.8.vec.insert6 = insertelement <2 x i64> %.0.vec.insert4, i64 %1, i32 1
  %2 = icmp eq <2 x i64> %.8.vec.insert, %.8.vec.insert6
  %3 = sext <2 x i1> %2 to <2 x i8>
  %4 = shufflevector <2 x i8> %3, <2 x i8> undef, <2 x i32> <i32 1, i32 undef>
  %5 = and <2 x i8> %4, %3
  %6 = extractelement <2 x i8> %5, i32 0
  ret i8 %6
}

attributes #0 = { noinline norecurse nounwind readnone uwtable "target-cpu"="skylake" }
```

Reply:
Can you check what `Vector<double>` used to generate?

Reply:
@filipnavara same codegen

Reply:
The existing code uses a vector of `i8` for each of the intermediate values (instead of a vector of integers with width equal to that of the input vector's element type), which results in bad generated code:
https://godbolt.org/z/3JNyZy

So we could instead use the type of `lhs` directly for vectors of integers, use `i64` for `double`, and use `i32` for float, and get reasonable code without needing to manually emit sequences of intrinsics.